### PR TITLE
fim watch_files and watch_new_files

### PR DESCRIPTION
**Pull request by jettero**
This roughly represents my thinking on the os.walk() and file-watches... I tested this trivially by hand, and it works, but I'm suspicious how often I see "add file-watch path=blah" in the debug logs. 

Does process() add watches too often? Does it rebuild them? I need to do a little more testing before I'd suggest merging this. I just wanted to get the PR up so we could talk about it.

(no merge please)

### CODE REVIEW

**Comment by jettero** [comment]
This definitely needs a couple fixes (at a minimum).

**Comment by jettero** [comment]
Ahh, do-not-merge label. I took "[WIP]" out of the name. (bitbucket habit)

**Comment by cedwards** [comment]
I like where you're going with this. Thanks, @jettero .

**Comment by jettero** [comment]
@basepi this branch is getting weird on me. It's trying to do too much, but it's helped figuring it all out too.

I have the following things going on:

* splitting out repetition into methods
* (re)tracking watched files/dirs in a new watchmanager class -- is this a welcome notion at all?
* managing the fs.inotify.max_user_watches (optionally by config)
* debugging the orphaned inotify watch problem -- this is my main goal right now, the above has been useful for this part

I'm assuming "yes,", but should I aim to split up all the things this branch is trying to do? or is some/all of it useful enough to just keep working on it as is? You're supposed to do a single feature in a feature branch. Fail.

**Comment by basepi** [comment]
I'm not too worried about doing all that work in one branch. At least it's all related to pulsar. ;)

If you can split it out, that could be useful, but it seems like a lot of this stuff would intertwine, making merge conflicts an issue once split out. Just keep them here, I think.

**Comment by jettero** [comment]
OK, great. That's a load off. It *is* intertwined.

**Comment by jettero** [comment]
I claim this is as far as it's going to get without running it for a few days and carefully observing problems.

File: hubblestack/extmods/modules/pulsar.py [hunk 1]
```diff
@@ -66,17 +66,529 @@ def _enqueue(revent):
...
+        to_set = __opts__.get('pulsar', {})
+
+        # Is there a better way to tell if __opts__ updated?
+        # Is it worth checking anyway? Seems only to come up in tests/
+        # todo?: attempt to re-read /etc/hubble/hubble sometimes?
+        counter = len( set(config).symmetric_difference( set(to_set) ) )
+        if counter == 0:
+            for k in config:
+                if config[k] != to_set[k]:
+                    counter += 1
```
Comment (basepi) [inline hubblestack/extmods/modules/pulsar.py]: We have no plans to reload __opts__ at this time. I would skip this to save processing time.

File: hubblestack/extmods/modules/pulsar.py [hunk 2]
```diff
@@ -66,17 +66,529 @@ def _enqueue(revent):
...
+        else:
+            the_list.append(str_wrapper(e))
+
+    # finally, wrap the whole decision set in a decision wrapper
+    def _final(val):
+        for i in the_list:
+            if i( val ):
+                return True
+        return False
+    time.sleep(2)
```
Comment (basepi) [inline hubblestack/extmods/modules/pulsar.py]: Why?

File: hubblestack/extmods/modules/pulsar.py (hunk 1 above)
Comment (jettero) [reply hubblestack/extmods/modules/pulsar.py]: I doubt it adds much processing time. ... I'll go ahead and remove it though.

File: hubblestack/extmods/modules/pulsar.py (hunk 2 above)
Comment (jettero) [reply hubblestack/extmods/modules/pulsar.py]: Oh, that's probably a stray debug thing. I thought I got them all ... they were marked with `# debug` so I could find them and get them out.

Comment (jettero) [reply hubblestack/extmods/modules/pulsar.py]: Oh, I should have said also: I intentionally didn't spend much time on docs. It's hard enough to get the docs right when you're sure you're looking at the final-ish form. I wasn't sure how much of this PR would get accepted, so I didn't spend a lot of time on it. I have notes on what to put where though.

Comment (jettero) [reply hubblestack/extmods/modules/pulsar.py]: lol, my test runs are SO FAST without that sleep(2) in there. how did I not notice that was in there all week?

**Comment by basepi** [comment]
As far as I'm concerned, once you remove the tabs that I noted above, this is ready to merge. Do you want to do docs on this PR or on a separate one?

File: hubblestack/extmods/modules/pulsar.py [hunk 3]
```diff
@@ -66,17 +66,523 @@ def _enqueue(revent):
...
+    def _listify_anything(cls, x, discard_none=True):
+        ''' _iterate_anything, then uniquify and force a list return; because,
+            pyinotify's __format_param, checks only isinstance(item,list)
+        '''
+        s = set( cls._iterate_anything(x, discard_none=discard_none) )
+        return list(s)
+
+    def _add_db(self, parent, **items):
+	# this assumes bijection, which isn't necessarily true
+	# (we hope it's true though)
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/extmods/modules/pulsar.py]: lol. how did that happen?

**Comment by jettero** [comment]
Oh, ... Yeah, why don't I do the docs as a separate PR.

**Pull request merged**